crates/biome_parser/CONTRIBUTING.md

```diff
@@ -0,0 +1,382 @@
+
+For example, `parse_for_statement` or `parse_expression`.
+
+### Signature
+Most parse rules take a `&mut` reference to the parser as their only parameter and return a `ParsedSyntax`.
+
+```rust,ignore
+fn parse_rule_name(&mut: Parser) -> ParsedSyntax {}
+```
+
+You're free to add additional parameters to your function if needed. There are rare cases where you want to consider returning `ConditionalParsedSyntax` as explained in [conditional syntax](#conditional-syntax)
+
+
+### Parsing a single node
+
+Let's assume you want to parse the JS `if` statement:
+
+```js
+JsIfStatement =
+ if
+ (
+ test: JsAnyExpression
+ )
+ consequent: JsBlockStatement
+ else_clause: JsElseClause?
+```
+
+#### Presence Test
+
+Now, the parsing function must first test if the parser is positioned at an `if` statement and return `Absent` if that's not the case.
+
+```rust, ignore
+if !p.at(T![if]) {
+ return ParsedSyntax::Absent;
+}
+```
+
+Why return `ParsedSyntax::Absent`? The function must return `ParsedSyntax::Absent` if the rule can't predict by the next token(s) if they form the expected node or not. Doing so allows the calling rule to decide if this is an error and perform an error recovery if necessary.  The second reason is to ensure that the rule doesn't return a node where all children are missing.
+
+Your rule implementation may want to consider more than just the first child to determine if it can parse at least some of the expected children.
+For example, the if statement rule could test if the parser is located at an `else` clause and then create an `if` statement where all children are missing except the `else` clause:
+
+```rust, ignore
+if !p.at(T![if]) && !p.at(T![else]){
+  return Absent
+}
+```
+
+Your implementation can also call into another parsing rule if the first child is a node and not a token.
+
+```rust, ignore
+let assignment_target = parse_assignment_target(p);
+
+if assignment_target.is_absent() {
+  return Absent;
+}
+
+let my_node = assignment_target.precede_or_missing();
+```
+
+But be careful with calling other rules. Your rule mustn't progress the parser - meaning that it can't
+advance in the parsing process and consume tokens - if it returns `Absent`.
+
+
+#### Parse children
+The parse rules will guide you in how to write your implementation and the parser infrastructure provides the following convenience APIs:
+
+* Optional token `'ident'?`: Use `p.eat(token)`. It eats the next token if it matches the passed-in token.
+* Required token `'ident'`: Use`p.expect(token)`. It eats the next token if it matches the passed-in token.
+It adds an `Expected 'x' but found 'y' instead` error and a missing marker if the token isn't present in the source code.
+* Optional node `body: JsBlockStatement?`: Use`parse_block_statement(p).or_missing(p)`. It parses the block if it is present in the source code and adds a missing marker if it isn't.
```

Comment:
Question: When is `or_missing()` actually necessary? I’ve found missing markers for optional nodes to be added automatically so far.

Reply:
Docs are **outdated** :D 

I just copied the content form `lib.rs`. Happy to review it later